logging/journald_core.go

```diff
@@ -124,3 +132,59 @@ func encodeJournaldFieldKey(key string) string {
 
 	return key
 }
+
+// visibleFieldsMsg creates a string to be appended to the log message including fields to be explicitly printed.
+//
+// When logging against journald, the zapcore.Fields are used as journald fields, resulting in not being shown in the
+// default journalctl output (short). While this is documented in our docs, missing error messages are usually confusing
+// for end users.
+//
+// This method takes an allow list (set, map of keys to empty struct) of key to be displayed - there is the global
+// variable journaldVisibleFields; parameter for testing - and a slice of zapcore.Fields, creating an output string of
+// the allowed fields prefixed by a whitespace separator. If there are no fields to be logged, the returned string is
+// empty. So the function output can be appended to the output message without further checks.
+func visibleFieldsMsg(visibleFieldKeys map[string]struct{}, fields []zapcore.Field) string {
+	if visibleFieldKeys == nil || fields == nil {
+		return ""
+	}
+
+	enc := zapcore.NewMapObjectEncoder()
+
+	for _, field := range fields {
+		if _, shouldLog := visibleFieldKeys[field.Key]; shouldLog {
+			field.AddTo(enc)
+		}
+	}
+
+	// The internal zapcore.encodeError function[^0] can result in multiple fields. For example, an error type
+	// implementing fmt.Formatter results in another "errorVerbose" field, containing the stack trace if the error was
+	// created by github.com/pkg/errors including a stack[^1].
+	//
+	// Therefore, the keys are checked again in the following loop before being inserted in the final array, having a
+	// capacity of twice the fields length to minimize the chance of a resize.
+	//
+	// [^0]: https://github.com/uber-go/zap/blob/v1.27.0/zapcore/error.go#L47
+	// [^1]: https://pkg.go.dev/github.com/pkg/errors@v0.9.1#WithStack
+	visibleFields := make([]string, 0, 2*len(enc.Fields))
+	for k, v := range utils.IterateOrderedMap(enc.Fields) {
+		if _, shouldLog := visibleFieldKeys[k]; !shouldLog {
+			continue
+		}
+
+		var fieldVerb string
+		switch v.(type) {
+		case rune, string, []byte, error:
+			fieldVerb = `%q`
```

Comment:
I doubt having `rune` in here is a good idea as it's just an [alias for `int32`](https://pkg.go.dev/builtin#rune). I think we're more likely to log an `int32` than a `rune` anywhere. That would also be worth a test case.

(I'm glad that Go is just as broken as C++ in this respect.[])

Reply:
Indeed. I have removed `rune` from the `%q` verb case. So far, I cannot think of a case where we have logged a rune[^1].

[^1]: I am not asking for a proof by contradiction :upside_down_face: